File: NIC.cpp
#include "NIC.h"
#include <iostream>
#include <cstring>

// Initialize static members
NIC* NIC::instance = nullptr;
bool NIC::instanceCreated = false;

// Private constructor
NIC::NIC() : manufacturer(""), macAddress(""), type("") {
    // Constructor is private, only called by getInstance()
}

// Destructor
NIC::~NIC() {
    // Clean up if this is the singleton instance
    if (instance == this) {
        instance = nullptr;
        instanceCreated = false;
    }
}

// Static method to get the single instance
NIC* NIC::getInstance() {
    if (!instanceCreated) {
        instance = new NIC();
        instanceCreated = true;
    }
    return instance;
}

// Initialize NIC data
void NIC::initializeNIC(const std::string& manufacturer, 
                       const std::string& macAddress, 
                       const std::string& type) {
    this->manufacturer = manufacturer;
    this->macAddress = macAddress;
    
    // Validate type - only allow Ethernet or Token Ring
    if (type == "Ethernet" || type == "Token Ring") {
        this->type = type;
    } else {
        this->type = "Unknown";
        std::cerr << "Warning: Invalid NIC type. Setting to 'Unknown'." << std::endl;
    }
}

// Getter methods
std::string NIC::getManufacturer() const {
    return manufacturer;
}

std::string NIC::getMACAddress() const {
    return macAddress;
}

std::string NIC::getType() const {
    return type;
}

// Display NIC information
void NIC::displayInfo() const {
    std::cout << "=== Network Interface Controller Information ===" << std::endl;
    std::cout << "Manufacturer: " << manufacturer << std::endl;
    std::cout << "MAC Address: " << macAddress << std::endl;
    std::cout << "Type: " << type << std::endl;
    std::cout << "=============================================" << std::endl;
}

// Check if NIC is initialized
bool NIC::isInitialized() const {
    return !manufacturer.empty() && !macAddress.empty() && !type.empty();
}

// Reset instance (for testing purposes)
void NIC::resetInstance() {
    if (instance) {
        delete instance;
        instance = nullptr;
        instanceCreated = false;
    }
}

File: NIC.cs
using System;

namespace FTPClientProject
{
    /// <summary>
    /// Network Interface Controller (NIC) class that represents the single network card on the machine.
    /// Implements singleton pattern to ensure only one instance can be created.
    /// </summary>
    public sealed class NIC
    {
        // Private static instance for singleton pattern
        private static readonly Lazy<NIC> _instance = new Lazy<NIC>(() => new NIC());
        
        // Private constructor to prevent external instantiation
        private NIC()
        {
            // Initialize with default values
            Manufacture = "Unknown";
            MacAddress = "00:00:00:00:00:00";
            Type = NetworkType.Ethernet;
        }
        
        /// <summary>
        /// Gets the single instance of the NIC class
        /// </summary>
        public static NIC Instance => _instance.Value;
        
        /// <summary>
        /// Gets or sets the manufacturer of the network card
        /// </summary>
        public string Manufacture { get; set; }
        
        /// <summary>
        /// Gets or sets the MAC address of the network card
        /// </summary>
        public string MacAddress { get; set; }
        
        /// <summary>
        /// Gets or sets the type of network connection
        /// </summary>
        public NetworkType Type { get; set; }
        
        /// <summary>
        /// Override ToString method to provide a readable representation of the NIC
        /// </summary>
        /// <returns>String representation of the NIC properties</returns>
        public override string ToString()
        {
            return $"NIC - Manufacturer: {Manufacture}, MAC: {MacAddress}, Type: {Type}";
        }
    }
    
    /// <summary>
    /// Enumeration for network connection types
    /// </summary>
    public enum NetworkType
    {
        /// <summary>
        /// Ethernet connection type
        /// </summary>
        Ethernet,
        
        /// <summary>
        /// Token Ring connection type
        /// </summary>
        TokenRing
    }
}